text/trace/0157-messaging-semantic-conventions.md

```diff
@@ -0,0 +1,184 @@
+
+### Consumer
+
+A "consumer" receives the message and acts upon it. It uses the context and
+data to execute some logic, which might lead to the occurrence of new events.
+
+The consumer receives, processes, and settles a message. "Receiving" is the
+process of obtaining a message from the intermediary, "processing" is the
+process of acting on the information a message contains, "settling" is the
+process of notifying an intermediary that a message was processed successfully.
+
+### Intermediary
+
+An "intermediary" receives a message to forward it to the next receiver, which
+might be another intermediary or a consumer.
+
+## Scenarios
+
+Producing and consuming a message involves five stages:
+
+```
+PRODUCER
+
+Create
+  |                            CONSUMER
+  v        +--------------+                   
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+       |
+                  ^               v
+                  .            Process
+                  .               |
+                  .               v
+                  . . . . . .  Settle
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed successfully.
+
+The semantic conventions need to define how to handle failures and retries in
+all stages that interface with the intermediary: publish, receive and settle.
+
+Based on this model, the following scenarios capture major requirements and
+can be used for prototyping, as examples, and as test cases.
+
+### Individual settlement
+
+Individual settlement systems imply independent logical message flows. A single
+message is created and published in the same context, and it's delivered,
+consumed, and settled as a single entity. Each message needs to be settled
+individually.
+
+Transport batching can be treated as a special case: messages can be
+transported together as an optimization, but are produced and consumed
+individually.
+
+```
++---------+ +---------+ +---------+ +---------+ +---------+ +---------+
+|Message A| |Message B| |Message C| |Message D| |Message E| |Message F|
++---------+ +---------+ +---------+ +---------+ +---------+ +---------+
+ Settled                 Settled                             Settled 
+```
+
+#### Examples
+
+1. RabbitMQ: TODO
+
+### Checkpoint-based settlement
+
+Checkpoint-based settlement systems are likely to consume multiple messages at
+once. Messages are created, published (possibly) in batches with other
+unrelated messages, and delivered in (reshuffled) batches.
+
+Messages are processed as a stream and settled to specific checkpoints. A
+checkpoint points to a position of the stream up to which messages were
+processed and settled. Messages cannot be settled individually, instead, the
+checkpoint needs to be forwarded.
```

Comment:
I don't think that's strictly true.

Though impacting performance, it is possible with Kafka to settle each message when the processing is complete

Reply:
"Individual settlement" means settling a message independent of the position of the message in the queue/stream.

In a Kafka message stream, you cannot settle message B before message A, if message B is at a position later in the stream (past A).

I'll try to make that more clear in the document.

Reply:
Thanks